server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -618,6 +618,38 @@ public Analyzer searchQuoteAnalyzer() {
         return this.searchQuoteAnalyzer;
     }
 
+    /**
+     * Builds a {@linkplain Function} to lookup mappers for a request, adding
+     * any {@code extraMapping} provided.
+     * @param extraMapping extra mappings parse and to add to the request
+     *        lookup or {@code null} if there aren't any extra mappings  
+     */
+    public Function<String, MappedFieldType> newFieldTypeLookup(Map<String, Object> extraMapping) {
+        if (extraMapping == null || extraMapping.size() == 0) {
+            return this::fieldType;
+        }
+        // TODO parsing an entire mapping is almost certainly "too big"
+        Mapping mapping = documentMapperParser().parseMapping("extra", extraMapping);
+        Map<String, MappedFieldType> extra = new HashMap<>();
+        for (Mapper m : mapping.root()) {
+            // TODO this should dig the sub fields out of object mappers. Probably.
+            for (MappedFieldType ft: m.convertToSearchTimeMappings()) {
+                MappedFieldType fromIndexMapping = fieldType(ft.name());
+                if (fromIndexMapping != null) {
+                    throw new IllegalArgumentException("No shadowing?! Was [" + fromIndexMapping + "] attempted to add [" + mapping + "]");
+                }
+                extra.put(ft.name(), ft);
+            }
+        }
+        return fullName -> {
+            MappedFieldType searchTime = extra.get(fullName);
```

Comment:
do we want to give the precedence to ephemeral runtime fields if they are defined with the same name as a mapped field? Or do we want to reject the search?

Reply:
I've been rejecting them at parse time. But I could see the argument to take preference.

Reply:
Rejecting is good, I did not realize that that is what you did.